### PR TITLE
WIP: estimated input latency audit

### LINKED ISSUE #28: metric: Thread availability
Capturing how available the main thread is. This is a proxy metric to for potential input latency, as we cannot predict when input will hit the browser.

------------------------------------
_Update_: See details at `Timestamp: Main thread ready for interaction` in #27
--------------------------------------------

##### Scoring

S-curve handled by Ben's math. Basic expectations:

* Goal: 90% percentile main thread task duration ~ under 300ms
* Goal: Maximum main thread task duration is < 400ms
* 100/100 is 90% percentile figure is <12ms
* 0/100 is 90% percentile figure is <1000ms.
* TBD: Determine curve.

### PR DESCRIPTION
**Pull request by brendankenny**
aka thread availability, Risk to Responsiveness, expected task queueing time, etc. Fixes #28

screen shot 2016-06-14 at 01 03 15

still a few things to take care of, like making scoring work and it needs to get in shape to also be useable for measuring TTI.

### CODE REVIEW

File: src/lib/traces/tracing-processor.js [hunk 1]
```diff
@@ -97,21 +97,122 @@ class TraceProcessor {
...
+   * selected percentiles within a window of the main thread.
+   * @param {!Array<!Object>} trace
+   * @param {number=} startTime Optional start time of range of interest. Defaults to trace start.
+   * @param {number=} endTime Optional end time of range of interest. Defaults to trace end.
+   * @return {!Array<{percentile: number, time: number}>}
+   */
+  static getRiskToResponsiveness(trace, startTime, endTime) {
+    // TODO(bckenny): filter for top level slices ourselves?
+    const tracingProcessor = new TraceProcessor();
+    const model = tracingProcessor.init(trace);
```
Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: Let's pass in this model and require the caller to do init() on their own.
that makes it more compatible with the way TTI will use this: [like this-ish](https://github.com/paulirish/lighthouse/commit/f256ac966337c08a589839fedc01c6f9f6845ced#diff-41dc490e500fbf48d1d60d848c4dcee2)

File: src/lib/traces/tracing-processor.js [hunk 2]
```diff
@@ -97,21 +97,122 @@ class TraceProcessor {
...
+    const startEvent = trace.find(event => {
+      return event.name === 'TracingStartedInPage';
+    });
+    const mainThread = TraceProcessor._findMainThreadFromIds(model, startEvent.pid, startEvent.tid);
+
+    // Find durations of all slices in range of interest.
+    const durations = [];
+    mainThread.sliceGroup.topLevelSlices.forEach(slice => {
+      // Discard slices outside range.
+      if (slice.end <= startTime || slice.start >= endTime) {
```
Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: there was some weird intersecting helper function. you decided to nuke it?

File: src/lib/traces/tracing-processor.js [hunk 3]
```diff
@@ -97,21 +97,122 @@ class TraceProcessor {
...
+
+  /**
+   * Calculates the maximum queueing time (in ms) of high priority tasks for
+   * selected percentiles within a window of the main thread.
+   * @param {!Array<!Object>} trace
+   * @param {number=} startTime Optional start time of range of interest. Defaults to trace start.
+   * @param {number=} endTime Optional end time of range of interest. Defaults to trace end.
+   * @return {!Array<{percentile: number, time: number}>}
+   */
+  static getRiskToResponsiveness(trace, startTime, endTime) {
```
Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: 90% of this function is about filtering to the right events. if so, maybe that should be broken into its own function to keep the name precise?

File: src/lib/traces/tracing-processor.js [hunk 4]
```diff
@@ -97,21 +97,122 @@ class TraceProcessor {
...
+    }
+
+    return results;
+  }
+
+  /**
+   * Calculates the maximum queueing time (in ms) of high priority tasks for
+   * selected percentiles within a window of the main thread.
+   * @param {!Array<!Object>} trace
+   * @param {number=} startTime Optional start time of range of interest. Defaults to trace start.
```
Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: do me a favor and denote what sort of timestamp is used?  high-res time with milliseconds or nanoseconds?

File: src/lib/traces/tracing-processor.js (hunk 1 above)
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: done

File: src/lib/traces/tracing-processor.js (hunk 2 above)
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: yeah, all it did was the first conditional here, and it seemed like overkill to make an `ExplicitRange` object just for that one check

Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: cool sgtm

File: src/lib/traces/tracing-processor.js (hunk 4 above)
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: done

Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: (ms)

File: src/lib/traces/tracing-processor.js (hunk 3 above)
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: yeah, makes sense. I'll take a look at that tomorrow

**Comment by paulirish** [comment]
Presentation notes from lewis/irish discussion:

* Let's score on 90th percentile.  
* Highlight 90th percentile.
* "You have a 9 in 10 chance of input latency being XXX or shorter."

**Comment by paulirish** [comment]
I'm halfway through rebasing this to master. It's a little involved, so let me take on the rest.

File: src/lib/traces/tracing-processor.js [hunk 5]
```diff
@@ -97,21 +97,119 @@ class TraceProcessor {
...
+    // Find durations of all slices in range of interest.
+    // TODO(bckenny): filter for top level slices ourselves?
+    const durations = [];
+    mainThread.sliceGroup.topLevelSlices.forEach(slice => {
+      // Discard slices outside range.
+      if (slice.end <= startTime || slice.start >= endTime) {
+        return;
+      }
+
+      // Clip any at edges of range.
```
Comment (paulirish) [inline src/lib/traces/tracing-processor.js]: We should not clip.

**Comment by paullewis** [comment]
This looks great overall. I think we need a meta bug for the UI, though, because we need to make sure it's (eventually) understandable by as many folks as possible.

**Comment by brendankenny** [comment]
> I'm halfway through rebasing this to master. It's a little involved, so let me take on the rest.

how's that going?

File: src/lib/traces/tracing-processor.js (hunk 5 above)
Comment (brendankenny) [inline src/lib/traces/tracing-processor.js]: we can easily not clip, but that can change the question we're asking out from under us.

e.g. you ask for a window of 500ms and theres a 150ms task starting near the end, you're now looking at an up to ~650ms window

**Comment by paulirish** [comment]
closed by #436

**Pull request closed without merging**